lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -175,21 +180,59 @@ class PipelineOutlineLowering : public PipelineLowering {
     return builder.getStringAttr("s" + std::to_string(stageIdx));
   }
 
+  // Helper class to manage grabbing the various inputs for stage modules.
+  struct PipelineStageMod {
+    PipelineStageMod(PipelineOutlineLowering &parent, Block *block,
+                     hw::HWModuleOp mod) {
+      size_t nStageInputArgs = parent.pipeline.getNumStageArguments(block);
+      inputs = mod.getArguments().take_front(nStageInputArgs);
+      extInputs = mod.getArguments().slice(
+          nStageInputArgs, parent.getStageExtInputs(block).size());
+      clock = mod.getArgument(mod.getNumArguments() - 2);
+      reset = mod.getArgument(mod.getNumArguments() - 1);
+    }
+
+    ValueRange inputs;
+    ValueRange extInputs;
+    Value clock;
+    Value reset;
+  };
+
+  // Iterate over the external inputs to the pipeline, and determine which
+  // stages actually reference them. This will be used to generate the stage
+  // module signatures.
+  void gatherExtInputsToStages() {
+    for (auto extIn : pipeline.getInnerExtInputs()) {
+      for (auto *user : extIn.getUsers())
+        stageExtInputs[user->getBlock()].insert(extIn);
+    }
+  }
+
   LogicalResult run() override {
     pipelineName =
         (parentModule.getName() + "_p" + std::to_string(pipelineID)).str();
 
     cloneConstantsToStages();
 
     // Build the top-level pipeline module.
-    pipelineMod =
-        buildPipelineLike(pipelineName, pipeline.getInputs().getTypes(),
-                          pipeline.getResults().getTypes());
+    pipelineMod = buildPipelineLike(
+        pipelineName, pipeline.getInputs().getTypes(),
+        pipeline.getExtInputs().getTypes(), pipeline.getResults().getTypes());
     pipelineClk = pipelineMod.getBody().front().getArgument(
         pipelineMod.getBody().front().getNumArguments() - 2);
     pipelineRst = pipelineMod.getBody().front().getArgument(
         pipelineMod.getBody().front().getNumArguments() - 1);
 
+    // Maintain a mapping between external inputs and their corresponding
+    // block argument in the top-level pipeline.
+    for (auto [extIn, barg] :
+         llvm::zip(pipeline.getInnerExtInputs(),
```

Comment:
nit: this expression is getting a bit too complicated. Can you break out some of the computation and store it in a descriptively-named variable?